infrastructure/src/main/java/org/corfudb/infrastructure/log/AbstractLogSegment.java

```diff
@@ -2,44 +2,110 @@
 
-    final long startAddress;
+    static String SEGMENT_FILE_SUFFIX = ".log";
+
+    final long segment;
+
+    @NonNull
+    final String filePath;
+
+    @NonNull
+    final StreamLogParams logParams;
 
     @NonNull
-    final String fileName;
+    FileChannel writeChannel;
 
     @NonNull
-    final FileChannel writeChannel;
+    FileChannel readChannel;
 
     @NonNull
-    final FileChannel readChannel;
+    FileChannel scanChannel;
 
     // TODO: persist revision in header.
-    // Used for checking equality between stream log and garbage log segment in recovery.
+    // Used for checking equality between stream log segment and garbage log segment in recovery.
     long compactRevision;
 
-    // public abstract LogData read(long address);
+    AbstractLogSegment(long segment, String fileName, StreamLogParams logParams) {
+        this.segment = segment;
+        this.filePath = fileName;
+        this.logParams = logParams;
 
-    public abstract void append(long address, LogData entry);
-
-    // public Iterator<LogData> iterator() {
-    //     throw new NotImplementedException();
-    // }
+        try {
+            this.writeChannel = getChannel(fileName, false);
+            this.readChannel = getChannel(fileName, true);
+        } catch (IOException ioe) {
+            log.error("Error opening file {}", fileName, ioe);
+            IOUtils.closeQuietly(writeChannel);
+            IOUtils.closeQuietly(readChannel);
+            throw new IllegalStateException(ioe);
+        }
+    }
 
+    /**
+     * Close the segment, releasing the resources if any.
+     */
     public abstract void close();
+
+    /**
+     * Opens the file channel to manipulate the file, creating a new file if
+     * it does not exist. If the readOnly parameter is set, the channel will
+     * be opened in read-only mode.
+     *
+     * @param filePath the path of the file to open
+     * @param readOnly if the file should be opened only for read
+     * @return the file channel to manipulate the specified file
+     */
+    private FileChannel getChannel(String filePath, boolean readOnly) throws IOException {
+        if (readOnly) {
+            if (!new File(filePath).exists()) {
+                throw new FileNotFoundException(filePath);
+            }
+            return FileChannel.open(
+                    FileSystems.getDefault().getPath(filePath),
+                    EnumSet.of(StandardOpenOption.READ)
+            );
+        }
+
+        try {
+            EnumSet<StandardOpenOption> options = EnumSet.of(
+                    StandardOpenOption.READ,
+                    StandardOpenOption.WRITE,
+                    StandardOpenOption.CREATE_NEW);
+            FileChannel channel = FileChannel.open(FileSystems.getDefault().getPath(filePath), options);
+
+            // First time creating this segment file, need to sync the parent directory
+            File segFile = new File(filePath);
+            syncDirectory(segFile.getParent());
```

Comment:
Can you plain why sync the parent directory?

Reply:
Because a newly created file might not be reflected in the parent directory. This is copy-pasted from current code.